File: src/medium/Q0230.cpp
// 230. Kth Smallest Element in a BST
// https://leetcode.com/problems/kth-smallest-element-in-a-bst/

#include <functional>
#include <vector>

using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
};

// Time: O(k) if k > log n, otherwise O(log n), where n is the number of nodes
// in the tree.
// Space: O(k)
auto kth_smallest(TreeNode *root, int k) -> int {
  auto vals = vector<int>();
  vals.reserve(k);

  function<void(const TreeNode *)> dfs = [&](const TreeNode *node) {
    if (node == nullptr) {
      return;
    }
    dfs(node->left);
    if (vals.size() < k) {
      vals.push_back(node->val);
    }
    if (vals.size() < k) {
      dfs(node->right);
    }
  };

  dfs(root);
  return vals.back();
}

File: src/medium/Q0230/Q0230.py
"""
230. Kth Smallest Element in a BST
https://leetcode.com/problems/kth-smallest-element-in-a-bst/
"""
from __future__ import annotations
from dataclasses import dataclass


@dataclass
class TreeNode:
    val: int
    left: TreeNode | None
    right: TreeNode | None


# Time: O(k) if k > log n, otherwise O(log n), where n is the number of nodes in the tree.
# Space: O(k)
def kth_smallest(root: TreeNode | None, k: int) -> int:
    vals = []

    def dfs(node: TreeNode | None) -> None:
        if node is None:
            return
        dfs(node.left)
        if len(vals) < k:
            vals.append(node.val)
        if len(vals) < k:
            dfs(node.right)

    dfs(root)
    return vals[-1]
